### PR TITLE
Null checks for arguments in MPI function calls

### PR DESCRIPTION
**Pull request by Bayyapu**

### CODE REVIEW

**Comment by Bayyapu** [comment]
test:jenkins/ch3/

**Comment by Bayyapu** [comment]
test:jenkins/ch3/

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
It looks like there is a test listed in `test/mpi/errors/comm/Makefile.am`, but the corresponding source file is not present in the repo. `comm_remote_size_test.c`. Can you make sure it gets added?

Another note: naming each test `<function>_test` is not very descriptive. I would like to suggest calling them `<function>_nullarg` instead.

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

File: test/mpi/errors/datatype/type_size_x_nullarg.c [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    MPI_Errhandler_set(MPI_COMM_WORLD, MPI_ERRORS_RETURN);
+
+    /* Checking type_size_x for NULL variable */
+    type = MPI_INT;
+    mpi_errno = MPI_Type_size_x(type, NULL);
+    MPI_Error_class(mpi_errno, &errclass);
+    if (errclass != MPI_ERR_ARG)
+        ++errs;
+
+    MPI_type_free(&type);
```
Comment (raffenet) [inline test/mpi/errors/datatype/type_size_x_nullarg.c]: typo in function name

File: test/mpi/errors/datatype/type_vector_nullarg.c [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    MPI_Type_commit(&type2);
+    /* Checking type_vector for NULL variable */
+    mpi_errno = MPI_Type_vector(3, 2, 3, type2, NULL);
+    MPI_Error_class(mpi_errno, &errclass);
+    if (errclass != MPI_ERR_ARG)
+        ++errs;
+
+    MPI_Type_free(&type);
+    MPI_Type_free(&type2);
+    MTest_Finlaize();
```
Comment (raffenet) [inline test/mpi/errors/datatype/type_vector_nullarg.c]: typo in function name

**Comment by raffenet** [comment]
@Bayyapu can you add a commit to this PR to make sure the `errors/datatype` directory is run by the testsuite? It appears `datatype` is missing from `test/mpi/errors/testlist.in`.

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

File: test/mpi/errors/datatype/getcnterr.c [hunk 3]
```diff
@@ -37,9 +37,9 @@ int main(int argc, char *argv[])
      */
     MPI_Errhandler_set(MPI_COMM_WORLD, MPI_ERRORS_RETURN);
     err = MPI_Get_elements(&recvstatus, type_ia, &count);
-    if (err == MPI_SUCCESS && count != MPI_UNDEFINED) {
+    if (err != MPI_SUCCESS && count == MPI_UNDEFINED) {
```
Comment (raffenet) [inline test/mpi/errors/datatype/getcnterr.c]: According to the comment above, this call should always return an error, so I don't think this new condition is correct. The old one looks OK to me.

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

File: test/mpi/errors/datatype/testlist [hunk 4]
```diff
@@ -1 +1,10 @@
-getcnterr 1
+getcnterr 1 xfail=ticket2697
```
Comment (raffenet) [inline test/mpi/errors/datatype/testlist]: Can you make this change to xfail `getcnterr` a separate commit? It is unrelated to the new tests being added in this commit.

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

File: test/mpi/errors/datatype/Makefile.am [hunk 5]
```diff
@@ -12,5 +12,14 @@ EXTRA_DIST = testlist
 ## for all programs that are just built from the single corresponding source
 ## file, we don't need per-target _SOURCES rules, automake will infer them
 ## correctly
-noinst_PROGRAMS = getcnterr
+noinst_PROGRAMS = getcnterr \
+                  type_contiguous_nullarg \
```
Comment (raffenet) [inline test/mpi/errors/datatype/Makefile.am]: The program names here do not match the names of the new testfiles.

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Pull request merged**